### PR TITLE
Add admin blackhole(s) shuttle event

### PR DESCRIPTION
**Pull request by timothymtorres**
## About The Pull Request
This adds a new **Black Hole** admin shuttle event while the escape shuttle is in transit.  The regular version spawns a small 1x1 singularity that falls through the ship.  There is also a special adminbus Kobayashi Maru version that spawns multiple black holes rapidly.

## Why It's Good For The Game
dreamseeker_MxqHmKXXQ8

Tell me this doesn't look fun?

## Changelog
:cl:
add: Add admin blackhole shuttle event with a normal version and suicidal version.
fix: Fix several shuttle event runtimes
/:cl:

### CODE REVIEW

**Comment by MrMelbert** [comment]
The shuttle really doesn't need random singulos thrown at it to be fun or chaotic, it already is, this just takes away from it.

**Comment by timothymtorres** [comment]
### The Orion Trail

_Sensors indicate that a black hole's gravitational field is affecting the region of space we were headed through. We could stay of course, but risk of being overcome by its gravity, or we could change course to go around, which will take longer._

---

> The shuttle really doesn't need random singulos thrown at it to be fun or chaotic, it already is, this just takes away from it.

You have selected option:

- [ ] `BUTTON_SPEED_PAST`
- [x] `BUTTON_GO_AROUND`

**Comment by Likteer** [comment]
No, it does not.

**Comment by carlarctg** [comment]
I meaaan a funny reference to Orion Trail where an option appears on the shuttle console to go around (adds 2 minutes to transit) or go through (40% chance for a vortex anomaly to slowly move past the shuttle, taking a small bit of the hull away at the edges) does sound fun

this is too much though

**Comment by Loafin34** [comment]
let singulo cook

**Comment by MrMelbert** [comment]
> I meaaan a funny reference to Orion Trail where an option appears on the shuttle console to go around (adds 2 minutes to transit) or go through (60% chance for a vortex anomaly to slowly move past the shuttle, taking a small bit of the hull away at the edges) does sound fun

This is a much preferred alternative, gives some more functionality to the shuttle pilot console too

**Comment by nukularpower** [comment]
> I meaaan a funny reference to Orion Trail where an option appears on the shuttle console to go around (adds 2 minutes to transit) or go through (60% chance for a vortex anomaly to slowly move past the shuttle, taking a small bit of the hull away at the edges) does sound fun
> 
this would be great, though i'd say just use a singulo rather than a weak ass vortex

**Comment by MrMelbert** [comment]
> This is a much preferred alternative, gives some more functionality to the shuttle pilot console too

Actually gonna double back on this and suggest that it grants you the option to *speed up* the shuttle at risk of singulo rather than making it slower if you decline

**Comment by MrMelbert** [comment]
Thought about it more and I'm gonna 1984 this can unavoidably happen. Something like above is fine though.

**Comment by timothymtorres** [comment]
Both are now only spawnable admin events.

**Comment by improvedname** [comment]
> ### The Orion Trail
> _Sensors indicate that a black hole's gravitational field is affecting the region of space we were headed through. We could stay of course, but risk of being overcome by its gravity, or we could change course to go around, which will take longer._
> 
> > The shuttle really doesn't need random singulos thrown at it to be fun or chaotic, it already is, this just takes away from it.
> 
> You have selected option:
> 
> * [ ]  `BUTTON_SPEED_PAST`
> * [x]  `BUTTON_GO_AROUND`

wish this was a option that would delay the shuttle timer but the shuttle travel is safe
and if the console is emagged it should speed trough it (increasing drastic shuttle events) as you fastened the shuttle takeoff as tradeoff

File: code/modules/power/singularity/singularity.dm [hunk 1]
```diff
@@ -475,3 +476,7 @@
 /obj/singularity/deadchat_controlled/Initialize(mapload, starting_energy)
 	. = ..()
 	deadchat_plays(mode = DEMOCRACY_MODE)
+
+// Special singularity that spawns for shuttle events only
```
Comment (Time-Green) [inline code/modules/power/singularity/singularity.dm:480]:
```suggestion
/// Special singularity that spawns for shuttle events only
```

**Review by Time-Green — APPROVED** [review]
Shuttle cling and shuttle event code is fine, I'm not really sure what the changes in _component are supposed to accomplish

**Comment by timothymtorres** [comment]
There was a runtime with hyperspace component that was triggering when the shuttle would get replaced via admin command.  It involved a hivebot that would not be registered to the components `parent` for some reason.  I did a little debugging but couldn't figure out why it was happening, so I just made this band-aid fix that checks to make sure `parent` is used.  Whatever is causing the hivebot to not register to a parent still exists. (maybe simple mobs aren't able to inherit certain components/traits)

File: code/modules/shuttle/shuttle_events/_shuttle_events.dm [hunk 2]
```diff
@@ -129,7 +129,7 @@
 	. = pick_weight(spawning_list)
 	if(remove_from_list_when_spawned) //if we have this enabled, we decrease the pickweight by 1 till it runs out
 		spawning_list[.] -= 1
-		if(spawning_list[.] < 1)
+		if(spawning_list[.] < 0)
```
Comment (Time-Green) [inline code/modules/shuttle/shuttle_events/_shuttle_events.dm:132]: Why below zero? Pickweight will never pick a value if it is set to 0, so -1 shouldn't occur

File: code/datums/components/shuttle_cling.dm [hunk 3]
```diff
@@ -97,8 +101,13 @@
 		return ALL_GOOD
 
 	if(!isliving(movee))
+		var/datum/component/singularity/singularity = movee.GetComponent(/datum/component/singularity)
+		if(singularity) // nothing can block the singularity
+			return SUPER_NOT_HOLDING_ON
```
Comment (Time-Green) [inline code/datums/components/shuttle_cling.dm:106]: Could you add a trait for this instead? I don't enjoy typechecks or GetComponent when they can be avoided

**Review by Time-Green — CHANGES_REQUESTED** [review]

File: code/datums/components/_component.dm [hunk 4]
```diff
@@ -128,22 +128,24 @@
...
 
-		else //just us
-			parents_components -= I
+			else //just us
+				parents_components -= I
 
-	if(!parents_components.len)
-		parent._datum_components = null
+		if(!parents_components.len)
+			parent._datum_components = null
```
Comment (Time-Green) [inline code/datums/components/_component.dm:148]: Could you revert these changes? It's because the hivebox is qdeling in initialize without passing the correct hints, I'm putting up a fix

Comment (timothymtorres) [reply code/datums/components/_component.dm]: Applied with https://github.com/tgstation/tgstation/pull/77188/commits/26ec41da8eb659dbd5d6a295f7656e898adbc563

File: code/datums/components/shuttle_cling.dm (hunk 3 above)
Comment (timothymtorres) [reply code/datums/components/shuttle_cling.dm:105]: Applied with https://github.com/tgstation/tgstation/pull/77188/commits/4cf8865d55796735fcdec391ff6b42039428a19a

File: code/modules/shuttle/shuttle_events/_shuttle_events.dm (hunk 2 above)
Comment (timothymtorres) [reply code/modules/shuttle/shuttle_events/_shuttle_events.dm:132]: Because `spawn_list.Remove(.)` removes the selected object (and deletes it?) before it's properly returned.

Comment (timothymtorres) [reply code/modules/shuttle/shuttle_events/_shuttle_events.dm]: Hmmm, very strange.  Testing it just now the previous `1` value now seems to work fine.  Maybe it was because I had set the list to be `spawning_list = list(/obj/singularity/shuttle_event)` instead of `spawning_list = list(/obj/singularity/shuttle_event = 1)` or playing with `self_destruct_when_empty`.

Reverted with https://github.com/tgstation/tgstation/pull/77188/commits/b69ed678202e8451985eaa81f8f211f7f10e28d9
